components/supervisor/pkg/supervisor/tasks.go

```diff
@@ -0,0 +1,234 @@
+// Copyright (c) 2020 TypeFox GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package supervisor
+
+import (
+	"bufio"
+	"context"
+	"io"
+	"io/ioutil"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/gitpod-io/gitpod/common-go/log"
+	csapi "github.com/gitpod-io/gitpod/content-service/api"
+	"github.com/gitpod-io/gitpod/supervisor/api"
+	"github.com/gitpod-io/gitpod/supervisor/pkg/backup"
+	"github.com/gitpod-io/gitpod/supervisor/pkg/terminal"
+)
+
+type tasksManager struct {
+	config            *Config
+	tasks             map[string]*api.TasksStatus
+	ready             chan struct{}
+	terminalService   *terminal.MuxTerminalService
+	inWorkspaceHelper *backup.InWorkspaceHelper
+}
+
+func newTasksManager(config *Config, terminalService *terminal.MuxTerminalService, inWorkspaceHelper *backup.InWorkspaceHelper) *tasksManager {
+	return &tasksManager{
+		config:            config,
+		terminalService:   terminalService,
+		inWorkspaceHelper: inWorkspaceHelper,
+		tasks:             make(map[string]*api.TasksStatus),
+		ready:             make(chan struct{}),
+	}
+}
+
+func (tm *tasksManager) Run(ctx context.Context, wg *sync.WaitGroup) {
+	defer wg.Done()
+	defer close(tm.ready)
+
+	tasks, err := tm.config.getGitpodTasks()
+	if err != nil {
+		log.WithError(err).Fatal()
+		return
+	}
+	if tasks == nil {
+		log.Info("no gitpod tasks found")
+		return
+	}
+
+	select {
+	case <-ctx.Done():
+		return
+	case <-tm.inWorkspaceHelper.ContentReady():
+		contentSource, _ := tm.inWorkspaceHelper.ContentSource()
+		headless := tm.config.GitpodHeadless != nil && *tm.config.GitpodHeadless == "true"
+		for i, task := range *tasks {
+			index := strconv.Itoa(i)
+			prebuildLogFileName := "/workspace/.gitpod/prebuild-log-" + index
+
+			var taskCommands []*string
+			if headless {
+				// prebuild
+				taskCommands = []*string{task.Before, task.Init, task.Prebuild}
+			} else if contentSource == csapi.WorkspaceInitFromPrebuild {
+				// prebuilt
+				legacyPrebuildLogFileName := "/workspace/.prebuild-log-" + index
+				printlogs := "[ -r " + legacyPrebuildLogFileName + " ] && cat " + legacyPrebuildLogFileName + "; [ -r " + prebuildLogFileName + " ] && cat " + prebuildLogFileName + "; true"
+				taskCommands = []*string{task.Before, &printlogs, task.Command}
+			} else if contentSource == csapi.WorkspaceInitFromBackup {
+				// restart
+				taskCommands = []*string{task.Before, task.Command}
+			} else {
+				// init
+				taskCommands = []*string{task.Before, task.Init, task.Command}
+			}
+			if taskCommands == nil {
+				continue
+			}
+			command := reduce(taskCommands, func(composed string, command string) string {
+				if composed != "" {
+					composed += " && "
+				}
+				return composed + "{\n" + command + "\n}"
+			})
```

Comment:
```
var cmds []string
for _, cmd := range taskCommands {
    if cmd == nil {
        continue
    }
    cmd = fmt.Sprintf("{ %s }", strings.TrimSpace(cmd))
    cmds = append(cmds, cmd)
}
command := strings.TrimSpace(strings.Join(cmds, " && "))
```

This would express the same "logic" without resorting to a "generic" reduce function.

Reply:
rewritten, not sure whether it is better, i still need a function to avoid code duplication below and using fmt requires allocating more variables compare to string concatenation